Fix all optimization-inhibiting integer-to-pointer casts

Svetlitski opened the pull request:
Following from PR #2481, we replace all integer-to-pointer casts [which hide pointer provenance information (and thus inhibit optimizations)](https://clang.llvm.org/extra/clang-tidy/checks/performance/no-int-to-ptr.html) with equivalent operations that preserve this information. I have enabled the corresponding clang-tidy check in our static analysis CI so that we do not get bitten by this again in the future.

interwq commented on include/jemalloc/internal/jemalloc_internal_decls.h:
```diff
@@ -105,4 +105,6 @@ isblank(int c) {
 #  undef small
 #endif
 
+typedef char byte_t;
```
nit: add some comments about why using this instead of uintptr_t etc

interwq commented on include/jemalloc/internal/jemalloc_internal_types.h:
```diff
@@ -99,7 +99,8 @@ typedef enum malloc_init_e malloc_init_t;
 #define ALIGNMENT_ADDR2BASE(a, alignment)				\
-	((void *)((uintptr_t)(a) & ((~(alignment)) + 1)))
+	((void *)(((byte_t *)(a)) - (((uintptr_t)(a)) -			\
+	    ((uintptr_t)(a) & ((~(alignment)) + 1)))))
```
I feel this can use some comments too (which should include the effective / original calculation). It's pretty counter intuitive why this would allow better codegen. Maybe also mention `ALIGNMENT_ADDR2CEILING` for their similarities.

interwq commented on src/background_thread.c:
```diff
@@ -368,6 +368,7 @@ check_background_thread_creation(tsd_t *tsd,
 		pre_reentrancy(tsd, NULL);
 		int err = background_thread_create_signals_masked(&info->thread,
+			/* NOLINTNEXTLINE(performance-no-int-to-ptr) */
 		    NULL, background_thread_entry, (void *)(uintptr_t)i);
```
Given how popular this pattern is, maybe add a wrapper like `void *int_to_ptr_nolint()` and put the comment inside?

Svetlitski replied:
If you want, sure. But I personally feel like the ugliness of this is a feature, not a bug. The lint (and the act of suppressing the lint) _should_ make you feel icky: smuggling random integers in pointers is something we should try to avoid when possible.

Svetlitski replied:
Never mind, I think you're right and this can be improved, I will address this.

interwq replied:
Sorry forgot to follow up here. I agree there should be some intentional ugliness, but perhaps in the form of a extra verbose function name, then within the wrapper function we can add detailed comments about why the hassle. Another factor is, some functions have a long list of args, and a wrapper would apply to the exact argument instead of the line.

Svetlitski replied:
How do you feel about meeting in the middle (see latest force-push)? I've extracted the vast majority of offending casts into macros (instead of using magic numbers), which dramatically cuts down the number of `NOLINTNEXTLINE` comments we need.

interwq replied:
This looks great. I didn't realize adding `NOLINTNEXTLINE` above the `#define`s would work for all references, but it certainly makes sense. Thanks for chasing down these magic numbers too!

Svetlitski committed:
Define `PROF_TCTX_SENTINEL` instead of using magic numbers

This makes the code more readable on its own, and also sets the stage
for more cleanly handling the pointer provenance lints in a following
commit.

Svetlitski committed:
Fix all optimization-inhibiting integer-to-pointer casts

Following from PR #2481, we replace all integer-to-pointer casts [which
hide pointer provenance information (and thus inhibit
optimizations)](https://clang.llvm.org/extra/clang-tidy/checks/performance/no-int-to-ptr.html)
with equivalent operations that preserve this information. I have
enabled the corresponding clang-tidy check in our static analysis CI so
that we do not get bitten by this again in the future.

Svetlitski committed:
Define `SBRK_INVALID` instead of using a magic number

Svetlitski committed:
Remove vestigial `TCACHE_STATE_*` macros
